### PR TITLE
Merge DDPStrategy and DDPSpawnStrategy in Lite

### LINKED ISSUE #14707: Merge Spawn Strategy in Lite
## 🚀 Feature

The DDPStrategy and DDPSpawnStrategy are two very similar strategies. Fundamentally, they only differ in the way processes are managed. Their logic can be merged and de-duplicated, since over time we have factored out all things related to process launching and environments to separate classes.

### Motivation

- Less maintenance of duplicated code
- Easier to do in Lite first, with the reduced complexity and can then be integrated in PL eventually, as PL becomes more dependent on Lite
- Easier to understand code for maintainers and users as well. 

### Pitch

Merge the two implementations (resulting in the class named DDPStrategy). Register the same class with different init args for launcher settings. 
Logical consequence: DDPShardedStrategy and DDPSpawnShardedStrategy will also merge. 

Note:
- The TPUSpawn Strategy should first be refactored to subclass ParallelStrategy directly. It's implementation is vastly different from DDPSpawnStrategy, and it has its own launcher anyway. 

### Alternatives

Leave as is.

### Additional context

Lite has now their own implementations
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @rohitgr7 @tchaton @carmocca @kaushikb11 @akihironitta

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #14707 

### Does your PR introduce any breaking changes? If yes, please list them.

Yes.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @awaelchli @carmocca

### CODE REVIEW

File: tests/tests_lite/test_connector.py [hunk 1]
```diff
@@ -435,18 +427,17 @@ def test_strategy_choice_sharded(strategy, expected_strategy, precision, expecte
     assert isinstance(connector.precision, expected_precision)
 
 
+# TODO: check if redundant
```
Comment (awaelchli) [inline tests/tests_lite/test_connector.py:430]: TODO

File: tests/tests_lite/test_connector.py [hunk 2]
```diff
@@ -397,23 +389,23 @@ def test_strategy_choice_cpu_str(strategy, strategy_class):
     assert isinstance(connector.strategy, strategy_class)
 
 
-@pytest.mark.parametrize("strategy_class", [DDPSpawnStrategy, DDPStrategy])
-def test_strategy_choice_cpu_instance(strategy_class):
-    connector = _Connector(strategy=strategy_class(), accelerator="cpu", devices=2)
-    assert isinstance(connector.strategy, strategy_class)
+# TODO: check if other test already does this
```
Comment (awaelchli) [inline tests/tests_lite/test_connector.py:392]: TODO

File: src/lightning_lite/strategies/xla.py [hunk 3]
```diff
@@ -38,7 +38,7 @@
     from torch_xla.distributed.parallel_loader import MpDeviceLoader
 
 
-class XLAStrategy(DDPSpawnStrategy):
+class XLAStrategy(DDPStrategy):  # TODO(lite)
```
Comment (awaelchli) [inline src/lightning_lite/strategies/xla.py:41]: TODO

File: src/lightning_lite/strategies/ddp.py [hunk 4]
```diff
@@ -100,17 +110,19 @@ def process_group_backend(self) -> Optional[str]:
 
     def _configure_launcher(self) -> None:
         assert self.cluster_environment is not None
-        if not self.cluster_environment.creates_processes_externally:
+        if self._start_method == "popen":
+            # if not self.cluster_environment.creates_processes_externally:
```
Comment (awaelchli) [inline src/lightning_lite/strategies/ddp.py:114]: TODO

File: src/lightning_lite/strategies/ddp.py [hunk 5]
```diff
@@ -53,6 +61,7 @@ def __init__(
         precision: Optional[Precision] = None,
         process_group_backend: Optional[str] = None,
         timeout: Optional[timedelta] = default_pg_timeout,
+        start_method: Literal["popen", "spawn", "fork", "forkserver"] = "popen",
```
Comment (awaelchli) [inline src/lightning_lite/strategies/ddp.py:64]: The three options from the spawn strategy and the popen from the normal ddp strategy are now all available to choose from here. The previous `DDPSpawnStrategy` becomes `DDPStrategy(start_method="spawn")`

File: src/lightning_lite/strategies/ddp.py [hunk 6]
```diff
@@ -100,17 +110,18 @@ def process_group_backend(self) -> Optional[str]:
 
     def _configure_launcher(self) -> None:
         assert self.cluster_environment is not None
-        if not self.cluster_environment.creates_processes_externally:
+        if self._start_method == "popen":
             self._launcher = _SubprocessScriptLauncher(self.cluster_environment, self.num_processes, self.num_nodes)
```
Comment (awaelchli) [inline src/lightning_lite/strategies/ddp.py:114]: The `if not self.cluster_environment.creates_processes_externally` is checked already inside `_SubprocessScriptLauncher`

File: src/lightning_lite/strategies/ddp.py [hunk 7]
```diff
@@ -142,6 +153,8 @@ def barrier(self, *args: Any, **kwargs: Any) -> None:
             torch.distributed.barrier()
 
     def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
+        if not _distributed_available():
+            return obj
```
Comment (awaelchli) [inline src/lightning_lite/strategies/ddp.py:157]: This was in the ddp spawn strategy. Merging the two broadcast implementation into one.

File: src/lightning_lite/strategies/launchers/multiprocessing.py [hunk 8]
```diff
@@ -121,7 +121,7 @@ def _wrapping_function(
     ) -> None:
         if global_states:
             global_states.restore()
-        self._strategy._local_rank = process_idx
+        os.environ["LOCAL_RANK"] = str(process_idx)
```
Comment (awaelchli) [inline src/lightning_lite/strategies/launchers/multiprocessing.py:124]: This is now the same way the subprocess launcher sets the local rank for the ddp strategy.

File: tests/tests_lite/test_connector.py [hunk 9]
```diff
@@ -397,23 +389,17 @@ def test_strategy_choice_cpu_str(strategy, strategy_class):
     assert isinstance(connector.strategy, strategy_class)
 
 
-@pytest.mark.parametrize("strategy_class", [DDPSpawnStrategy, DDPStrategy])
-def test_strategy_choice_cpu_instance(strategy_class):
-    connector = _Connector(strategy=strategy_class(), accelerator="cpu", devices=2)
-    assert isinstance(connector.strategy, strategy_class)
```
Comment (awaelchli) [inline tests/tests_lite/test_connector.py:403]: redundant now. the case is already covered by the one after the next test.

File: src/lightning_lite/strategies/ddp.py [hunk 10]
```diff
@@ -150,20 +163,36 @@ def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
...
         )
-        strategy_registry.register(
-            "ddp",
-            cls,
-            description=cls.__class__.__name__,
+        for name, start_method in entries:
+            strategy_registry.register(
+                name,
+                cls,
+                description=f"DDP strategy with `start_method` '{start_method}'",
```
Comment (carmocca) [inline src/lightning_lite/strategies/ddp.py:176]:
```suggestion
                description=f"DDP strategy with `start_method={start_method!r}`",
```

File: src/lightning_lite/strategies/ddp.py [hunk 11]
```diff
@@ -150,20 +163,36 @@ def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
...
+            ("ddp_find_unused_parameters_false", "popen"),
+            ("ddp_spawn_find_unused_parameters_false", "spawn"),
+            ("ddp_fork_find_unused_parameters_false", "fork"),
+            ("ddp_notebook_find_unused_parameters_false", "fork"),
         )
+        for name, start_method in entries:
+            strategy_registry.register(
+                name,
+                cls,
+                description=f"DDP strategy with `find_unused_parameters` as False and `start_method` '{start_method}'",
```
Comment (carmocca) [inline src/lightning_lite/strategies/ddp.py:190]:
```suggestion
                description=f"DDP strategy with `find_unused_parameters` as False and `start_method={start_method!r}`",
```

**Review by carmocca — APPROVED** [review]
Great!

File: src/lightning_lite/strategies/ddp.py (hunk 10 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/ddp.py:176]: Then one could directly use `f"{start_method=}"`

Comment (carmocca) [reply src/lightning_lite/strategies/ddp.py:176]: Thats a 3.8+ feature

**Review by justusschock — APPROVED** [review]
